compiler/optimizer/OMRSimplifierHandlers.cpp

```diff
@@ -5260,6 +5260,118 @@ static bool skipRemLowering(int64_t divisor, TR::Simplifier *s)
    }
 
 
+/**
+ * @brief Reduce Ifcmp Greater-Than-Or-Equal to Ifcmp Equal in a specific case.
+ *
+ * If
+* - the first child is an "and" node
+* - the second child of the "and" node is a constant N
+* - the second child of the ifcmpge node is a constant N,
+* Then, we can reduce ifcmpge to ifcmpeq.
+*
+* For example, convert
+* ificmpge --> block_5
+*  iand
+*    iload  <...>
+*    iconst 16
+*  iconst 16
+*
+*  into
+*
+* ificmpeq --> block_5
+*  iand
+*    iload  <...>
+*    iconst 16
+*  iconst 16
+*
+* Since the constants are equal, the case of greater-than is impossible,
+* and we can reduce ifcmpge to ifcmpeq. In other words, the result of iand
+* cannot be greater than the constant, and we can simply perform an equality check,
+* which provides further optimization opportunities.
+*/
+class IfcmpgeReducer
+   {
+   public:
+   IfcmpgeReducer(TR::Node *node, TR::Simplifier *s) : _node(node), _s(s) {}
+
+   /**
+    * @brief Get a specific ifcmpeq opcode based on a ifcmpge opcode.
+    *
+    * Each ifcmpge opcode is mapped to its corresponding ifcmpeq opcode.
+    * E.g., ifiucmpge should be substituted with ifiucmpeq.
+    *
+    * @param node An ifcmpge node
+    * @return An ifcmpeq node
+    */
+   static TR::ILOpCodes getNewOpCode(TR::Node *node)
+      {
+      switch (node->getOpCodeValue())
+          {
+          case TR::ifbcmpge:
+             return TR::ifbcmpeq;
+          case TR::ifbucmpge:
+             return TR::ifbucmpeq;
+          case TR::ifscmpge:
+             return TR::ifscmpeq;
+          case TR::ifsucmpge:
+             return TR::ifsucmpeq;
+          case TR::ificmpge:
+             return TR::ificmpeq;
+          case TR::ifiucmpge:
+             return TR::ifiucmpeq;
+          case TR::iflcmpge:
+             return TR::iflcmpeq;
+          case TR::iflucmpge:
+             return TR::iflucmpeq;
+          defaul:
+             TR_ASSERT(false, "Ifcmp reduction is not implemented for node [%p]\n", node);
+          }
+      }
+
+   /**
+    * @brief Checks whether an ifcmpge node can be reduced to ifcmpeq.
+    *
+    * @return True if the reduction is possible.
+    */
+   template<typename T>
+   bool reducible() const
+      {
+      TR::Node * firstChild = _node->getFirstChild(), * secondChild = _node->getSecondChild();
+
+      if (firstChild->getOpCode().isAnd() &&
+         firstChild->getSecondChild()->getOpCode().isIntegralConst() &&
+         secondChild->getOpCode().isIntegralConst())
+         {
+         T constAnd = firstChild->getSecondChild()->getConst<T>();
+         T constIfcmp = secondChild->getConst<T>();
+
+         if (constAnd == constIfcmp)
+            return true;
```

Comment:
This needs to also check that either the comparison is unsigned, or the mask has the high/sign bit clear.  Otherwise it's possible for the result of `iand` to compare greater.  For example, consider the following tree:

    ificmpge
      iand
        foo
        iconst -2
      iconst -2

If `foo` is non-negative, then the result of `iand` will also be non-negative, and therefore > -2.  Slightly more aggressively, it should be possible to transform this tree in cases where `foo` is known to be negative.

Reply:
Nice catch @jdmpapin. @Leonardo2718 surely this should have been caught by the unit tests as `std::vector<T> TR::Test::const_values()` which is used by @last5bits returns all combinations including `negative_value<T>` which was -42 in this case. Any idea why we don't have a failing unit test here?

Reply:
On second thought presumably the following test would have been generated:

```
(method return="NoType"
  (block name="b" fallthrough="u"
    (istore temp="val"
       (iconst 1) )
      (ificmpge target="t
          (iand"
              (iload temp="val")
              (iconst -42))
          (iconst -42) ) )
  (block name="u" (return))
  (block name="t" (return))
```

However I don't see us comparing the result of the test against a valid output. So there is a failing test case here but we are just not verifying it.

Reply:
I think this may be a good enough reason to justify calling the compiled body.

Reply:
What we really need to do for these tests is to modify the TRIL generated code to be something like this (using above as example):

```
(method return="Int32"
  (block name="b" fallthrough="u"
    (istore temp="val"
       (iconst 1) )
      (ificmpge target="t
          (iand"
              (iload temp="val")
              (iconst -42))
          (iconst -42) ) )
  (block name="u" (ireturn (iconst 1)))
  (block name="t" (ireturn (iconst 0)))
```

And test this method with:

```
ASSERT_EQ(0, compiler.compile()) << "Compilation failed unexpectedly\n" << "Input trees: " << inputTrees;

auto entry_point = compiler.getEntryPoint<int32_t (*)(void)>();
volatile auto exp = param.oracle(param.lhs, param.rhs);
volatile auto act = entry_point();
ASSERT_EQ(exp, act);
```

Where `param.orcale` is constructed using the lambda:

```
[]() { return ((1 & -42) >= 1); }
```

Reply:
Seems reasonable to me @fjeremic.

There is some room for simplifying things here (compared to what's in `ArithmeticTest.cpp`) because we're only dealing with one opcode.

> We should also add several more constants in `TRTest::const_values()`.

Definitely. This will likely help other tests as well.

Reply:
I don't know what optimizations are applied when compiling from tril, but the test above would just constant fold the entire branch if local CSE were allowed to run before simplifier.  It may be a good idea to change `val` into a parameter, to ensure that it's completely opaque

Reply:
@jdmpapin 

Tril uses JitBuilder as a backend so it will run whatever optimization strategy is used there. However, this specific test is explicitly run with `disableGlobalVP` and `disableLocalCSE`. So, that deals with part of the problem.

> It may be a good idea to change val into a parameter, to ensure that it's completely opaque.

Definitely a worthwhile option.

Ideally, I would prefer to test both cases (when val is a parameter and when val is constant), but that's an improvement for later.

Reply:
Fixed all of this in bb54ec4